lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -530,6 +533,108 @@ def internal_check_future_reserved_word(o, name)
     end
   end
 
+  def internal_check_file_namespace(o, name, file)
+    return if file.nil? || file == '' #e.g. puppet apply -e '...'
+
+    lc_file = file.downcase
+
+    file_namespace = @file_to_namespace[lc_file]
+    if file_namespace.nil?
+      return if @file_to_namespace.key?(lc_file)
+
+      file_namespace = @file_to_namespace[lc_file] = namespace_for_file(lc_file)
+      return if file_namespace.nil?
+    end
+
+    if !name.downcase.start_with?(file_namespace)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    end
+  end
+
+  NEVER_MATCH = '\b'.freeze
+
+  def namespace_for_file(file)
+    path = Pathname.new(file)
+
+    return nil if path.extname != ".pp"
+    path = path.expand_path
+
+    env = Puppet.lookup(:current_environment)
+    return nil if env.nil?
+
+    return nil if initial_manifest?(path, env)
+
+    #All auto-loaded files from modules come from a module search path dir
+    relative_path = get_module_relative_path(path, env)
+
+    return nil if relative_path.nil?
+
+    #If a file comes from a module, but isn't in the right place, always error
+    names = dir_to_names(relative_path)
+    return NEVER_MATCH if names.nil?
+
+    names.join("::").freeze
+  end
+
+  def initial_manifest?(path, env)
+    manifest_setting = env.nil? ? nil : env.manifest
+
+    # If there is no manifest, all files should be checked for correct location/namespace
+    if manifest_setting.nil? || manifest_setting == :no_manifest
+      return false
+    end
+
+    #Smashing case for comparison - is this correct in all environments?
+    manifest_setting = Pathname.new(manifest_setting).expand_path.to_s.downcase
+    full_path = path.expand_path.to_s.downcase
+
+    full_path == manifest_setting || full_path.start_with?(manifest_setting)
+  end
+
+  def get_module_relative_path(dir, environment)
+    return nil if dir.nil? || environment.nil?
+    file_path = Pathname.new(dir)
+    paths = environment.full_modulepath.map { |p| Pathname.new(p) }
+
+    parent_path = paths.select { |env_dir| is_parent_dir_of(env_dir, file_path) }.first
+    return nil if parent_path.nil?
+
+    file_path.relative_path_from(parent_path).to_s.split(File::SEPARATOR)
+  end
+
+  def is_parent_dir_of(parent_dir, child_dir)
+    return false if parent_dir.class != Pathname || child_dir.class != Pathname
```

Comment:
Rather than comparing classes here, perhaps you could do an is_a?, i.e.
```
return false unless parent_dir.is_a?(Pathname) && child_dir.is_a?(Pathname)
```
or if you really want explicit class comparison:
```
return false unless parent_dir.instance_of?(Pathname) && child_dir.instance_of?(Pathname)
```